lib/charms/saml_integrator/v0/saml.py

```diff
@@ -300,4 +300,7 @@ def update_relation_data(self, relation: ops.Relation, saml_data: SamlRelationDa
             relation: the relation for which to update the data.
             saml_data: a SamlRelationData instance wrapping the data to be updated.
         """
-        relation.data[self.charm.model.app].update(saml_data.to_relation_data())
+        # Update only if the relation data has changed to prevent triggering unnecessary events
+        relation_data = saml_data.to_relation_data()
+        if relation.data[self.charm.model.app] != relation_data:
```

Comment:
What type is the relation data here? Depending on the type the comparison operator won't necessarily do what you think. E.g., if it is a custom type, unless the `__eq__` method has been implemented it will just compare pointer locations. The inbuilt types should be ok, e.g., if it is a dict then it should work. For example, see the repl:

```
{"key": "value"} == {"key": "value"}
True
>>> {"key": object()} == {"key": object()}
False
```

Reply:
It's a Dict[str, str]